nav2_astar_planner/src/astar_planner.cpp

```diff
@@ -40,33 +56,475 @@ AStarPlanner::execute(const nav2_tasks::ComputePathToPoseCommand::SharedPtr comm
+
+  planner_->setCostmap(&costmapData[0], true, allow_unknown_);
+
+  unsigned int mx, my;
+  if (!worldToMap(world_point.x, world_point.y, mx, my)) {
+    return false;
+  }
+
+  int map_start[2];
+  map_start[0] = 0;
+  map_start[1] = 0;
+
+  int map_goal[2];
+  map_goal[0] = mx;
+  map_goal[1] = my;
+
+  planner_->setStart(map_start);
+  planner_->setGoal(map_goal);
+
+  return planner_->calcNavFnAstar();
+}
+
+bool
+AStarPlanner::getPlanFromPotential(
+  const geometry_msgs::msg::Pose & goal,
+  nav2_msgs::msg::Path & plan)
+{
+  // clear the plan, just in case
+  plan.poses.clear();
+
+  // Goal should be in global frame
+  double wx = goal.position.x;
+  double wy = goal.position.y;
+
+  // the potential has already been computed, so we won't update our copy of the costmap
+  unsigned int mx, my;
+  if (!worldToMap(wx, wy, mx, my)) {
+    RCLCPP_WARN(
+      get_logger(),
+      "The goal sent to the navfn planner is off the global costmap."
+      " Planning will always fail to this goal.");
+    return false;
+  }
+
+  int map_goal[2];
+  map_goal[0] = mx;
+  map_goal[1] = my;
+
+  planner_->setStart(map_goal);
+
+  planner_->calcPath(costmap_.metadata.size_x * 4);
+
+  // extract the plan
+  float * x = planner_->getPathX();
+  float * y = planner_->getPathY();
+  int len = planner_->getPathLen();
+
+  plan.header.stamp = this->now();
+  plan.header.frame_id = global_frame_;
+
+  for (int i = len - 1; i >= 0; --i) {
+    // convert the plan to world coordinates
+    double world_x, world_y;
+    mapToWorld(x[i], y[i], world_x, world_y);
+
+    geometry_msgs::msg::Pose pose;
+    pose.position.x = world_x;
+    pose.position.y = world_y;
+    pose.position.z = 0.0;
+    pose.orientation.x = 0.0;
+    pose.orientation.y = 0.0;
+    pose.orientation.z = 0.0;
+    pose.orientation.w = 1.0;
+    plan.poses.push_back(pose);
+  }
+
+  return !plan.poses.empty();
+}
+
+double
+AStarPlanner::getPointPotential(const geometry_msgs::msg::Point & world_point)
+{
+  unsigned int mx, my;
+  if (!worldToMap(world_point.x, world_point.y, mx, my)) {
+    return std::numeric_limits<double>::max();
+  }
+
+  unsigned int index = my * planner_->nx + mx;
+  return planner_->potarr[index];
+}
+
+bool
+AStarPlanner::validPointPotential(const geometry_msgs::msg::Point & world_point)
+{
+  return validPointPotential(world_point, default_tolerance_);
+}
+
+bool
+AStarPlanner::validPointPotential(
+  const geometry_msgs::msg::Point & world_point, double tolerance)
+{
+  double resolution = costmap_.metadata.resolution;
+
+  geometry_msgs::msg::Point p = world_point;
+  p.y = world_point.y - tolerance;
+
+  while (p.y <= world_point.y + tolerance) {
+    p.x = world_point.x - tolerance;
+    while (p.x <= world_point.x + tolerance) {
+      double potential = getPointPotential(p);
+      if (potential < POT_HIGH) {
+        return true;
+      }
+      p.x += resolution;
+    }
+    p.y += resolution;
+  }
+
+  return false;
+}
+
+
```

Comment:
There seems to be extra whitespace here.

Reply:
done